### ISSUE TITLE
Let `Memory` store arbitrary data for custom widgets.

### ISSUE THREAD

**Issue by optozorax** [emilk/egui#255]
**Is your feature request related to a problem? Please describe.**
I have this widget:
![image](https://user-images.githubusercontent.com/18114381/112620341-fabfb280-8e5a-11eb-9303-ef3922f23caa.png) ![image](https://user-images.githubusercontent.com/18114381/112620358-00b59380-8e5b-11eb-96f4-8ccd1581e46a.png)

This is the editor where the result can be previewed. And this widget has two states: `View` or `Edit`, this can be represented as `bool`. I believe this is more convenient to store this data in `egui::Memory` because this whole thing looks like some complicated widget that receives `&mut String`, and using this we can have the same benefits as other widgets have (able to serialize UI memory between runs). Also, creating custom memory and drag it into every function is quite inconvenient.

I looking for the code of `CollapsingHeader`, because it has some bool state too (opened or not). And found that `Memory` [hard-coded](https://github.com/emilk/egui/blob/master/egui/src/memory.rs#L30) all data for collapsing headers and collapsing header [uses this](https://github.com/emilk/egui/blob/master/egui/src/containers/collapsing_header.rs#L27) private access. And, as I see there is no public interface to store data for a custom widget.

**Describe the solution you'd like**
First, we can provide an interface to store just a single `bool`, `i32`, `String` using enums for our id.

But a more scalable approach is using `HashMap` with `Box<dyn Any + ...>`. Using it, we can store arbitrary types for our `id`. And default widgets can be painlessly rewritten using it.

We also can require `Serialize` + `Deserialize` traits for all objects that should be stored in this type-map (under feature, of course).

But this type-map can possibly add visible runtime overhead comparing to current hard-coded maps.

**Describe alternatives you've considered**
Not allowing a user to store data in `egui::Memory`.

**Additional context**
I want to try to implement this feature by myself and open a Pull Request. What do you think?

**Comment by optozorax** [comment]
So, I tried to implement described idea with `Box<dyn Any + 'static>`, the result can be found here: https://github.com/optozorax/egui/commit/7db30a00b6f0c64c8058310418eec8f024672e36. 

Current observations:
- Using `Box<dyn Any>` is very ugly. Even cloning done by miracle ugly hacks.
- This works until we need Serialization and Deserialization. Serialization is in some way possible using ugly hacks (storing `String` and using `serde_json`), but Deserialization is impossible. (maybe possible, but very very ugly and unsafe https://github.com/alecmocatta/serde_traitobject)
- Or, if we want to have this type of Serialization and Deserialization, we should make a very ugly user API.

So, the idea with any-map is failed, so I suggest it's better to provide the user a way to store their data using simple Rust enums, maybe also introducing helper functions/macros to get/set data from that enums, or parsing using `if let`.

**Comment by optozorax** [comment]
I tryed to sleep, but this idea came: deserialization can be deffered, and can be performed when we call get function with specialized type.

**Comment by optozorax** [comment]
So, I created a [draft PR](https://github.com/emilk/egui/pull/257) where you can view how the approach with `Any` may look like. And made `CollapsingHeader` using this. This seems to work.

**Pros:**
- Nice interface for a user.

**Cons:**
- Ugly and complicated internal code.
- After removing all `unwrap`s error handling will become very complicated.
- Some deserialization errors can be determined only after some time using the interface.
- New dependency: `serde_json` (for `persistence` feature).
- Now we can't reset individual types (because it can be loaded from a file, and stored in `ToDeserialize` state), only everything can be reset.
- Individual types can't be counted (same reason).

---

Another approach is introduce something like this type:
```rust
enum DataStorage {
  None,
  Bool(bool),
  Int(i64),
  Float(f32),
  String(String),
  Tuple(Box<DataStorage>, Box<DataStorage>)
}
```

**Pros:**
- Nice internal code without ugly hacks.
- Straightforward cloning, serialization, and deserialization.
- No extra dependencies.

**Cons:**
- Ugly interface for a user, now they have to parse this struct by themselves using `match` and `if let` every time.
- Some deserialization errors can be determined only after some time using the interface too because a user can expect something like `Tuple(Float(_), Bool(_))`, but now there is stored `Tuple(Float(_), Tuple(Bool(_), Bool(_)))` because a version of their code changed between runs.
- Now we also can't reset individual types (because we can't determine if type `Bool(_)` refers to our internal widget or to user `Bool(_)`), only everything can be reset.
- Individual types can't be counted (same reason).

**Comment by optozorax** [comment]
I said:
> After removing all `unwrap`s error handling will become very complicated.

But this is not quite true. Why user should care if some internal _temporary_ UI data is not saved correctly between runs of different versions of his program? I think a user should use only functions `get_or_insert(Id, T) -> &T`, `get_or_default(Id) -> &T`, and `insert(Id, T)` and that's all. They can't return `Option<>` or can't panic because they will insert provided/default value if some error happens. With this approach and this interface we can stop bothering about error origin: downcast error, deferred deserializing error, we can just insert the correct new/default value.

But this is just an idea. And possibly this can create some obscure and hard-to-find errors for complicated interfaces.

**Comment by optozorax** [comment]
I believe this feature can be done in two PRs:
1. Add API to `Memory` to store arbitrary data. Rewrite system/default widgets and containers to use this API.
2. Add an example on how to use this API. Maybe view/edit widget from the first message should suit.

Default widgets should be rewritten to use this API because users may rely on the source code of these widgets.

(I still want to implement this if you give approval)

**Comment by optozorax** [comment]
The downcast error may occur when one `Id` is used for different widgets. Maybe this error should be logged or provided to a user as `Result<>` :thinking:

**Comment by emilk** [comment]
Wow!

I went through exactly this thought process many months ago, but I never managed to crack the nut - but it seems you did!

Postponing deserialization until we know what type to deserialize to is really clever.

As for errors: this storage should only be used for things that are non-critical ("the selected tab" etc), so I think silently ignoring errors is fine (*). Of course we need to be very clear documenting this.


---- 

(*) it may also be about time that `egui` got some way to report errors. Either using the `log` crate, or using some sort of message system (part of the `Output` struct), but that is a topic for another issue.

**Comment by emilk** [comment]
As for JSON: I use it elsewhere, but I am considering switching it out for something that supports `f32::INFINITY` etc. Maybe [ron](https://github.com/ron-rs/ron)

**Comment by parasyte** [comment]
Using `ron` is quite a great idea. You can use [`ron::value::Map`](https://docs.rs/ron/0.6.4/ron/value/struct.Map.html) directly to store the state without building any bespoke dynamic types. I was considering [`json::object::Object`](https://docs.rs/json/0.12.4/json/object/struct.Object.html) but if you don't have any hard dependency on JSON as a serialization format, then there are loads of other alternatives to consider, and `ron` is a top contender.

Also, it's worth pointing out that `serde` supports [transcoding](https://docs.serde.rs/serde_transcode/index.html) between formats, so `ron` values can be transcoded to JSON for persistence.

**Comment by optozorax** [comment]
I started to implement an example widget about how to use `Memory` in custom widgets.

![image](https://user-images.githubusercontent.com/18114381/114277921-7f175580-9a57-11eb-96b0-3da86c6f811a.png)

Now I need to write a bunch of comments and wait for https://github.com/emilk/egui/pull/257 to be merged.

**Comment by emilk** [comment]
I just used this for the first time in https://github.com/emilk/egui/commit/daf2e1323839d0796ffac2df57e97044a218e21f - I think this is a good example of something that does not belong in the user state, and so egui can help you keep track of instead.
